Added the feature engineering with Feast tutorial (#1195)
hazelcast/hz-docs: docs/modules/integrate/pages/feature-engineering-with-feast.adoc

= Feature Engineering with Feast
:description: This tutorial will get you started with feature engineering using the Hazelcast integration with Feast.

> 

Author: Updated at: 0fbd59166130220da5208cc35582a26546185b92

                        stmt.setTimestamp(3, Timestamp.from(Instant.ofEpochSecond(utc.getFeatureTimestamp())));
                    }));

        return pipeline;
    }

    public static void backfillFeatures(HazelcastInstance hz, String dataSetPath, LocalDateTime earliestEndDate) {
        var endDate = earliestEndDate;
        for (int i = 0; i < 8; i++) {
            hz.getJet().newJob(createPipeline(dataSetPath, endDate));
            endDate = endDate.minusDays(1);
        }
    }

    public static void main(String[] args) {
        if (args.length == 0) {
            throw new RuntimeException("dataSetPath is required");
        }
        var hz = Hazelcast.bootstrappedInstance();
        var endDate = LocalDateTime.now();
        var dataSetPath = args[0];
        backfillFeatures(hz, dataSetPath, endDate);
    }
}
----

. You have to compile the Java code that creates the Jet jobs.
We provided an easy-to-use script to do that from inside the demo container:
+
[source,shell]
----
run build_jet batch_features
----

. You can now create the Jet jobs and run them:
[source,shell]
----
clc job submit build/jet/batch_features/libs/*.jar /home/hazelcast/data
----

> . You can list the running jobs and verify that the jobs completed successfully using the following command:

Author: Updated at: 0fbd59166130220da5208cc35582a26546185b92

[source,shell]
----
clc job list

> ----

Author: Updated at: 0fbd59166130220da5208cc35582a26546185b92

Outputs:
+
[source,output]
----
------------------------------------------------------------------------------------
 Job ID              | Name | Status    | Submitted           | Completed
------------------------------------------------------------------------------------
 0c0d-c9a3-c14d-0001 | N/A  | COMPLETED | 2024-07-24 19:15:19 | 2024-07-24 19:15:19
 0c0d-c9a3-c14b-0001 | N/A  | COMPLETED | 2024-07-24 19:15:17 | 2024-07-24 19:15:17
 0c0d-c9a3-c149-0001 | N/A  | COMPLETED | 2024-07-24 19:15:15 | 2024-07-24 19:15:15
 0c0d-c9a3-c147-0001 | N/A  | COMPLETED | 2024-07-24 19:15:13 | 2024-07-24 19:15:13
 0c0d-c9a3-c145-0001 | N/A  | COMPLETED | 2024-07-24 19:15:11 | 2024-07-24 19:15:11
 0c0d-c9a3-c143-0001 | N/A  | COMPLETED | 2024-07-24 19:15:09 | 2024-07-24 19:15:09
 0c0d-c9a3-c141-0001 | N/A  | COMPLETED | 2024-07-24 19:15:07 | 2024-07-24 19:15:07
 0c0d-c9a3-c140-0001 | N/A  | COMPLETED | 2024-07-24 19:15:05 | 2024-07-24 19:15:06
----

== Materialization

Materialization is the process of transferring features from the offline store to the online store. In this case, from PostgreSQL to Hazelcast.

> . Run the following command to materialize the features:

Author: Updated at: 0fbd59166130220da5208cc35582a26546185b92

[source,shell]
----
feast -c feature_repo materialize-incremental "2024-07-24T08:00:00"
----

. Running the command above created an IMap that corresponds to the "user_transaction_count_7d" feature in the Hazelcasst cluster.
You can list it using the following command:
+
[source,shell]
----
clc object list map

> ----

Reviewer: ```suggestion
----
+
```

Author: Updated at: 0fbd59166130220da5208cc35582a26546185b92

Outputs:
+
[source,output]
----
---------------------------------------
 Object Name
---------------------------------------
 feast_batch_user_transaction_count_7d
---------------------------------------
    OK Returned 1 row(s).
----

. Check the contents of the feature IMap to check the data written by Feast:
+
[source,shell]
----
clc map -n feast_batch_user_transaction_count_7d entry-set | head -10
----

== Summary

> In this tutorial, you learned how to setup a feature engineering project that uses Hazelcast as the online store and PostgreSQL as the offline store..

Reviewer: ```suggestion
In this tutorial, you learned how to set up a feature engineering project that uses Hazelcast as the online store and PostgreSQL as the offline store..
```

You also learned how to write Jet jobs that transform data and store it in a PostgreSQL table to be used by the Feast offline store.

== See Also

There is more to feature engineering with Hazelcast.

Check out our documentation on Feast at: TODO lINK.

> If you have any questions, suggestions, or feedback please do not hesitate to reach out to us via https://slack.hazelcast.com/[Hazelcast Community Slack].

Reviewer: ```suggestion
If you have any questions, suggestions, or feedback please do not hesitate to reach out to us through https://slack.hazelcast.com/[Hazelcast Community Slack].
```

. Run the following command to materialize the features:
+
[source,shell]
----
feast -c feature_repo materialize-incremental "2024-07-24T08:00:00"
----

. Running the command above created an IMap that corresponds to the "user_transaction_count_7d" feature in the Hazelcast cluster.
You can list it using the following command:
+
[source,shell]
----
clc object list map
----
+
Outputs:
+
[source,output]
----
---------------------------------------
 Object Name
---------------------------------------
 feast_batch_user_transaction_count_7d
---------------------------------------
    OK Returned 1 row(s).
----

. Check the contents of the feature IMap to check the data written by Feast:
+
[source,shell]
----
clc map -n feast_batch_user_transaction_count_7d entry-set | head -10
----

== Summary

In this tutorial, you learned how to set up a feature engineering project that uses Hazelcast as the online store and PostgreSQL as the offline store..
You also learned how to write Jet jobs that transform data and store it in a PostgreSQL table to be used by the Feast offline store.

== See Also

There is more to feature engineering with Hazelcast.

> Check out the our documentation about Feast:

Reviewer: ```suggestion
Check out our documentation about Feast:
```

Author: Updated at: https://github.com/hazelcast/hz-docs/pull/1195/commits/932db89934f4adb68537f9d97cb610171d9d9b6e
